### PR TITLE
Color correction

### PR DESCRIPTION
**Pull request by AEtHeLsYn**
With this commit I added color correction to Hyperion. There are two correction steps, RGB LED correction and RGB color correction, same as Arduino library. Also, option to perform HSL correction is added.

### CODE REVIEW

**Comment by penfold42** [comment]
I think this is good - is it documented ?

**Comment by redPanther** [comment]
thx for the code. Your PR had merge conflicts, could you resolve the conflicts?

**Comment by brindosch** [comment]
First of all
Thank you for your contribution!
@redPanther I appreciate your judge ;)
Could you provide the changed segment @AEtHeLsYn  .conf

Could you publish a link to your compiled bins? Would be good that "everyone" is able to test this in production.

OT: We should thought about how to handle the "testing" stage better. Any idea is welcome.

**Comment by AEtHeLsYn** [comment]
I'm already at work, after a couple of hours will provide the new .conf segment. I've this mod working for two months, I'll upload a completed binary to test it.

**Comment by brindosch** [comment]
Thank you very much for your efforts!

**Comment by AEtHeLsYn** [comment]
Ok, conflicts solved, nothing to worry about. About the changed segment in the conf. here it is:

`"correction" :
		[
			{
				"id"   : "default",
				"leds" : "*",
				"correctionValues" :
				{
					"red"   : 195,
					"green" : 215,
					"blue"  : 255
				}
			}
		],
		"temperature" :
		[
			{
				"id"   : "default",
				"leds" : "*",
				"correctionValues" :
				{
					"red"   : 255,
					"green" : 255,
					"blue"  : 255
				}
			}
		],
		"transform" :
		[`

As you can see I kept the color structure. The color and temperature correction perform the same operation, but I split them just to help to work with them.

Same with hsl transform: 
`"hsv" :
				{
					"saturationGain" : 1.0000,
					"valueGain"      : 1.0000
				},
				"hsl" :
				{
					"saturationGain" : 0.9500,
					"luminanceGain"  : 1.2000
				},`

**Comment by brindosch** [comment]
I have to admit, i am at the "understanding" stage.

Tha values
`				"red" :
				{
					"threshold"  : 0.0000,
					"gamma"      : 1.0000,
					"blacklevel" : 0.0000,
					"whitelevel" : 1.0000
				},
				"green" :
				{
					"threshold"  : 0.0000,
					"gamma"      : 1.0000,
					"blacklevel" : 0.0000,
					"whitelevel" : 1.0000
				},
				"blue" :
				{
					"threshold"  : 0.0000,
					"gamma"      : 1.0000,
					"blacklevel" : 0.0000,
					"whitelevel" : 1.0000
				}`
Are still there?

**Comment by AEtHeLsYn** [comment]
Yes, of course. I added possibilities to color management keeping the initial config.

**Comment by brindosch** [comment]
You mean in long-term this values are no longer needed?

**Comment by AEtHeLsYn** [comment]
No, you still need those values as I only added color correction options. The modifications I made are the same as FastLED library: [https://github.com/FastLED/FastLED/wiki/FastLED-Color-Correction](url)

But, after the raw RGB colors are corrected, you still can modify threshold, gamma, blacklevel and whitelevel of each channel. 

For example, you can apply temperature correction by, let's say [255,240,190]. Doing that, every B value gets corrected by B*(190/255). After that, you can still correct gamma or blacklevel.

**Comment by brindosch** [comment]
Sound very good! Thank you @AEtHeLsYn for explanation.
For easy use in HyperCon and space saving, i would "link" als three to one field at the GUI
The advanced users should stick with the .conf editing
`        "id"   : "default",
            "leds" : "*",`
Should be ok.

**Comment by AEtHeLsYn** [comment]
Ok, full commit now tested, everything is working fine. I still have to get used to git ;)

I did not change any code of hyperion-remote to be able to accept new values, I'll look into that.

**Comment by AEtHeLsYn** [comment]
Ok, added json messages integration as well as updated hyperion-remote, everything is working fine.

**Comment by AEtHeLsYn** [comment]
Compiled bins:

[https://www.dropbox.com/s/9nzzv3dsnflva08/bin.zip?dl=0](https://www.dropbox.com/s/9nzzv3dsnflva08/bin.zip?dl=0)

**Comment by Xruchai86** [comment]
Hello. I have backup my old bin folder and replace it with your´s. Then i edit my config with your "options". After all, hyperion won´t start the service and my led´s are off :(. Do i anything fals? sry for my english ^^''

**Comment by AEtHeLsYn** [comment]
What exactly did you change? You need to transform this part of config:

	"color" :
	{
		"transform" :
		[
			{
				"id"   : "default",
				"leds" : "*",
				"hsv" :
				{
					"saturationGain" : 1.0000,
					"valueGain"      : 1.0000
				},
				"red" :
				{
					"threshold"  : 0.0000,
					"gamma"      : 1.0000,
					"blacklevel" : 0.0000,
					"whitelevel" : 1.0000
				},
				"green" :
				{
					"threshold"  : 0.0000,
					"gamma"      : 1.0000,
					"blacklevel" : 0.0000,
					"whitelevel" : 1.0000
				},
				"blue" :
				{
					"threshold"  : 0.0000,
					"gamma"      : 1.0000,
					"blacklevel" : 0.0000,
					"whitelevel" : 1.0000
				}
			}
		],
		"smoothing" :
		{
			"type"            : "linear",
			"time_ms"         : 200,
			"updateFrequency" : 20.0000,
			"updateDelay"     : 0
		}
	},

Into this:

	"color" :
	{
		"correction" :
		[
			{
				"id"   : "default",
				"leds" : "*",
				"correctionValues" :
				{
					"red"   : 255,
					"green" : 255,
					"blue"  : 255
				}
			}
		],
		"temperature" :
		[
			{
				"id"   : "default",
				"leds" : "*",
				"correctionValues" :
				{
					"red"   : 255,
					"green" : 255,
					"blue"  : 255
				}
			}
		],
		"transform" :
		[
			{
				"id"   : "default",
				"leds" : "*",
				"hsv" :
				{
					"saturationGain" : 1.0000,
					"valueGain"      : 1.0000
				},
				"hsl" :
				{
					"saturationGain" : 1.0000,
					"luminanceGain"  : 1.0000
				},
				"red" :
				{
					"threshold"  : 0.0000,
					"gamma"      : 1.0000,
					"blacklevel" : 0.0000,
					"whitelevel" : 1.0000
				},
				"green" :
				{
					"threshold"  : 0.0000,
					"gamma"      : 1.0000,
					"blacklevel" : 0.0000,
					"whitelevel" : 1.0000
				},
				"blue" :
				{
					"threshold"  : 0.0000,
					"gamma"      : 1.0000,
					"blacklevel" : 0.0000,
					"whitelevel" : 1.0000
				}
			}
		],
		"smoothing" :
		{
			"type"            : "linear",
			"time_ms"         : 200,
			"updateFrequency" : 20.0000,
			"updateDelay"     : 0
		}
	},

**Comment by Xruchai86** [comment]
Okay thx :D. I used exactly your code but it won´t start (with or without the changes). When i use the old bin folder it work oO...My LED´s are extrem blue (in my normal config the whitelevel is 0,15 and green 0,25). I love your "mod" and i hope it find the way into the update very soon! Because i don´t know why it not work for me, i must wait. Thx for your fast answer :)

**Comment by brindosch** [comment]
Sorry for the delay, we can´t merge it until we tested this (and the users)
@Xruchai86 
It would be good, if you could find your problem. Testers are welcome

**Comment by Xruchai86** [comment]
Tomorrow i have more Time to try it. If I find the error, I will inform you.

**Comment by AEtHeLsYn** [comment]
Does someone else have problems with compiled binaries? I've been running the mod for almost two months.
@Xruchai86  it's useful to have some feedback to see what's happening there. Can you upload your full config file?

**Comment by Xruchai86** [comment]
sure! ehm...sry but can you say me where i can upload this files? i don´t have dropbox^^''

**Comment by AEtHeLsYn** [comment]
Copy & paste the config file using [http://pastebin.ca/](http://pastebin.ca/)

**Comment by Xruchai86** [comment]
This one is my [Default-Config](http://pastebin.ca/3402164) 
and this one the [Color-Correction](http://pastebin.ca/3402166)

**Comment by AEtHeLsYn** [comment]
Ok, thank you. Just to be sure, what hardware platform and operating system are you using to run hyperion?

**Comment by Xruchai86** [comment]
I use Raspbian Jessie (latest version) on a PI2 with WS2801 LED´s. I use  LogiLink Audio und Video Grabber with Fushicai USBTV007 Chip too.

**Comment by AEtHeLsYn** [comment]
@Xruchai86 I just made a test by downloading the bin folder and using your config file. On my side everything is working fine. Anyway you probably need to give execute permissions to the binaries that you downloaded, I had to do that too.

**Comment by Xruchai86** [comment]
ahhh that was it! Thank you, all work fine with my default config. Tomorrow i play a little with the new settings :D

**Comment by AEtHeLsYn** [comment]
@Xruchai86 I'll be grateful if you can post some feedback after you play a bit with the settings ;D

**Comment by Xruchai86** [comment]
So after 10 min...i am very happy with the results! I only change temperature and gamma! I don´t know how i "calibrate" it right with this option ( a little how to were nice)but i do it the old way and only touch this 2 settings and have a very nice effect.:D

**Comment by AEtHeLsYn** [comment]
If you take a look at FastLED link I posted before you can find all the info, it is pretty well documented. Anyway the first thing you should do is use the correction config to achieve a "perfect" while, that way you can equalize all color outputs. Then, you can play with temperature correction in case you want to get LED warmer or colder. After the above two steps you can adjust everything as before, but it is preferable to use hsv OR hsl correction, not both.
In summary, first step is color correction, then color temp, and last is color transform. That is the way the correction is achieved.

**Comment by brindosch** [comment]
@AEtHeLsYn 
I "reviewed" your code and have some question now.
1. Please delete the typo in .conf on line 85.
2. Is there a reason, why you split the remote values?
`IntParameter       & argCorrR      = parameters.add<IntParameter>      ('1', "correctionR"  , "Specify the red channel correction");`
I would prefer a solution that follows the present implementation like this
```
-g "0.95 1.0 1.0"
```
In your case:
```
-y "167 64 210"
-z "127 159 6"
```
I am sure that all available remotes just "spam" send the whole transform package and not split it up so there is no real benefit, just more code/commands to deal with (in apps) and it breaks with the present implementations.
Thank you for all efforts you invested in this :+1: 
I am proud to announce that HyperCon is ready for your pull request (90%)

**Comment by AEtHeLsYn** [comment]
@brindosch 
I won't be able to change the code until sunday, but no problems at all ;) 
I thought about converting RGB values into double between 0.0 and 1.0, but in my opinion there is a lot more online documentation that talks about correcting RBG using absolute values

**Comment by brindosch** [comment]
I am sure that people could better deal with 0-255 instead 0-1 :)
unbenannt
as a little spoiler :)

**Comment by AEtHeLsYn** [comment]
Well, i like more the new behavior. First, get white balance by correcting channel "strenght", then you can adjust everything else. Whitelvl does almost the same, but after gamma correction and at the same time that blacklvl. In my opinion, when you are dealing with color modifications, it's better to keep it simple (there is no need for temp correction, it's the same operation, but easier to control).
I would like to completely replace hsv with hsl and get rid of whitelvl inside the transform, but not now ;)

**Comment by AEtHeLsYn** [comment]
@brindosch 
Ok, protocol changed, everything working ok ;)

**Comment by brindosch** [comment]
@AEtHeLsYn 
Great!
`StringParameter    & argIdC        = parameters.add<StringParameter>   ('y', "qualifier" , "Identifier(qualifier) `
Is for submit the led index?
But i could still just send without the qualifier?
`-Y "R G B" `

**Comment by AEtHeLsYn** [comment]
Yes, exactly like the transform. You can submit -Y "R G B" or -Z "R G B" without the qualifier.

**Comment by brindosch** [comment]
@AEtHeLsYn 
I have problems with the remote.
the following is working
`hyperion-remote -s 1.0 -v 1.0 -u 1.0 -m 1.0 -g "1.0 1.0 1.0" -t "0.0 0.0 0.0" -b "0.0 0.0 0.0" -w "1.0 1.0 1.0"`
if i add -Y -Z. 
`hyperion-remote -s 1.0 -v 1.0 -u 1.0 -m 1.0 -g "1.0 1.0 1.0" -t "0.0 0.0 0.0" -b "0.0 0.0 0.0" -w "1.0 1.0 1.0" -Y "255 255 255" -Z "255 255 255"`
I get the following
```
Multiple commands found. Provide exactly one of the following options:
	  -c, --color <arg>       
	  -i, --image <arg>       
 	  -e, --effect <arg>      
```
and no values (all) are set

Thank you

**Comment by AEtHeLsYn** [comment]
@brindosch 

In fact, the remote is working as expected if u take a look at the code: 


`// check that exactly one command was given`
`int commandCount = count({argColor.isSet(), argImage.isSet(), argEffect.isSet(), argServerInfo.isSet(), argClear.isSet(), argClearAll.isSet(), colorTransform, argCorrection.isSet(), argTemperature.isSet()});`
`if (commandCount != 1)`


As you can see, only one command is allowed. As you are setting only transformation values (saturation, gamma...) the remote allows it as you are only sending one command (transform). When you set correction commands, you are trying to send more than one command.

**Comment by brindosch** [comment]
This is a problem in usage. Is there a reason why !=1 is chosen?
This means we never can´t set a transform, a correction and a temperature at the same time. (App usage to adjust the settings)

**Comment by AEtHeLsYn** [comment]
@brindosch 
I can take a look into modding hyperion-remote code to allow more than one command at the same time, shouldn't be difficult. I'll keep you posted
